internal/parser/parser.go

```diff
@@ -1061,50 +1062,122 @@ func (p *parser) parsePropertyList() []*ast.Property {
 		case token.STRING:
 			param = p.parseStringProperty()
 		default:
-			// TODO(jsternberg): BadExpression.
-			p.consume()
-			continue
+			param = p.parseInvalidProperty()
 		}
 		params = append(params, param)
-		if _, tok, _ := p.peek(); tok == token.COMMA {
-			p.consume()
+
+		if p.more() {
+			if _, tok, lit := p.peek(); tok != token.COMMA {
+				perrs = append(perrs, ast.Error{
+					Msg: fmt.Sprintf("expected comma in property list, got %s (%q)", tok, lit),
+				})
+			} else {
+				p.consume()
+			}
 		}
 	}
+	p.errs = append(p.errs, perrs...)
 	return params
 }
 
 func (p *parser) parseStringProperty() *ast.Property {
 	key := p.parseStringLiteral()
 	p.expect(token.COLON)
-	val := p.parseExpression()
+	val := p.parsePropertyValue()
+	var endPos ast.Position
+	if val != nil {
+		endPos = locEnd(val)
+	} else {
+		endPos = locEnd(key)
+	}
 	return &ast.Property{
 		Key:   key,
 		Value: val,
 		BaseNode: p.baseNode(p.sourceLocation(
 			locStart(key),
-			locEnd(val),
+			endPos,
 		)),
 	}
 }
 
 func (p *parser) parseIdentProperty() *ast.Property {
+	property := &ast.Property{}
+	var startPos, endPos ast.Position
+
 	key := p.parseIdentifier()
-	loc := key.Location()
-	property := &ast.Property{
-		Key:      key,
-		BaseNode: p.baseNode(&loc),
-	}
+	property.Key = key
+	startPos = locStart(key)
+	endPos = locEnd(key)
+
 	if _, tok, _ := p.peek(); tok == token.COLON {
 		p.consume()
-		property.Value = p.parseExpression()
-		property.Loc = p.sourceLocation(
-			locStart(key),
-			locEnd(property.Value),
-		)
+		val := p.parsePropertyValue()
+		if val != nil {
+			property.Value = val
+			endPos = locEnd(val)
```

Comment:
Here's another place where it's ok to just call `locEnd` on the null value.

I'm thinking I want to refactor how locations are created entirely so I would like the problem to be consistent so it is easy to find.